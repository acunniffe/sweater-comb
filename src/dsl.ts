import {
  ApiCheckDsl,
  EntityRule,
  Result,
  DocsLinkHelper,
  newDocsLinkHelper,
  runCheck,
} from "@useoptic/api-checks";
import {
  ConceptualLocation,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiOperationFact,
  IChange,
  IFact,
  OpenApiFieldFact,
  jsonPointerHelper,
  ILocation,
  OpenAPIV3,
} from "@useoptic/openapi-utilities";
import { genericEntityRuleImpl } from "@useoptic/api-checks/build/sdk/generic-entity-rule-impl";
import { ShouldOrMust } from "@useoptic/api-checks/build/sdk/types";
import { OpenApiResponseFact } from "@useoptic/openapi-utilities/build/openapi3/implementations/openapi3/openapi-traverser";

type SnykStablity = "wip" | "experimental" | "beta" | "ga";
type DateString = string; // YYYY-mm-dd
type ResourceName = string;

export interface SynkApiCheckContext {
  // Vervet provides context about the change itself. Since
  // Optic is analyzing two OpenAPI spec files, we need to tell it
  // when the change is supposed to happen, and the resource/version info
  // determined by the file's location in a directory structure.
  changeDate: DateString; // when the change did (or would, if proposed) occur
  changeResource: ResourceName; // the spec resource being changed
  changeVersion: {
    // the spec version being changed
    date: DateString;
    stability: SnykStablity;
  };

  // Vervet provides a mapping that indicates the resource version deprecation.
  // It has this information because it processes the entire source tree of
  // spec files.
  resourceVersions: {
    [ResourceName: string]: {
      // changeResource used to match this
      [DateString: string]: {
        // changeVersion.date used to match this
        [SnykStablity: string]: {
          // changeVersion.stability matches this
          deprecatedBy: {
            // the spec version that deprecates this one (if any) or null
            date: DateString;
            stability: SnykStablity; // could be higher stability than changed!
          } | null;
        };
      };
    };
  };
}

export interface SnykEntityRule<T, A>
  extends EntityRule<T, ConceptualLocation, SynkApiCheckContext, A> {}

export class SnykApiCheckDsl implements ApiCheckDsl {
  private checks: Promise<Result>[] = [];

  constructor(
    private nextFacts: IFact<any>[],
    private changelog: IChange<any>[],
    private nextJsonLike: OpenAPIV3.Document,
    private providedContext: SynkApiCheckContext
  ) {}

  checkPromises() {
    return this.checks;
  }

  getContext(location: ILocation): ConceptualLocation & SynkApiCheckContext {
    return {
      ...location.conceptualLocation,
      ...this.providedContext,
    };
  }

  get operations() {
    return genericEntityRuleImpl<
      OpenApiOperationFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.OperationObject
    >(
      OpenApiKind.Operation,
      this.changelog,
      this.nextFacts,
      (opFact) => `${opFact.method.toUpperCase()} ${opFact.pathPattern}`,
      (location) => this.getContext(location),
      (...items) => this.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(this.nextJsonLike, pointer)
    );
  }

  get context() {
    const change: IChange<SnykApiCheckDsl> = {
      location: {
        conceptualLocation: { path: "This Specification", method: "" },
        jsonPath: "/",
        conceptualPath: [],
        kind: "API",
      },
    };

    const value: ShouldOrMust<
      (
        context: SynkApiCheckContext,
        docs: DocsLinkHelper
      ) => Promise<void> | void
    > = {
      must: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          true,
          () => handler(this.providedContext, docsHelper)
        );
      },
      should: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          false,
          () => handler(this.providedContext, docsHelper)
        );
      },
    };

    return value;
  }

  get responses() {
    const dsl = this;

    return {
      ...genericEntityRuleImpl<
        OpenApiResponseFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.ResponsesObject
      >(
        OpenApiKind.Response,
        dsl.changelog,
        dsl.nextFacts,
        (response) => `response ${response.statusCode}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
      headers: genericEntityRuleImpl<
        OpenApiHeaderFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.HeaderObject
      >(
        OpenApiKind.ResponseHeader,
        dsl.changelog,
        dsl.nextFacts,
        (header) => `response header ${header.name}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
    };
  }

  get checkApiContext(): ShouldOrMust<
    (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
  > {
    const contextChangedHandler: (must: boolean) => ContextChangedRule["must"] =
      (must: boolean) => {
        return (statement, handler) => {
          const docsHelper = newDocsLinkHelper();
          const syntheticChange = {
            added: this.providedContext,
            location: {
              jsonPath: "/",
              conceptualPath: [],
              conceptualLocation: {
                path: "Entire Resource",
                method: "",
              },
              kind: "ContextRule",
            },
          };
          return runCheck(
            syntheticChange,
            docsHelper,
            "api lifeycle: ",
            statement,
            must,
            () => handler(this.providedContext, docsHelper)
          );
        };
      };

    return {
      must: contextChangedHandler(true),
      should: contextChangedHandler(false),
    };
  }

  get bodyProperties(): SnykEntityRule<
    OpenApiFieldFact,
    OpenAPIV3.SchemaObject
  > {
    const dsl = this;
    return genericEntityRuleImpl<
      OpenApiFieldFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.SchemaObject
    >(
      OpenApiKind.Field,
      dsl.changelog,
      dsl.nextFacts,
      (field) => `field ${field.key}`,
      (location) => dsl.getContext(location),
      (...items) => dsl.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
    );
  }
}

type ContextChangedRule = ShouldOrMust<
  (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
>;
