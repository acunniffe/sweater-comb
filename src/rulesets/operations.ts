import { SnykApiCheckDsl } from '../dsl';
const { expect } = require('chai');

export const rules = {
  operationId: ({ operations, responses }: SnykApiCheckDsl) => {
    operations.requirement.must(
      'have the correct operationId format',
      (operation) => {
        expect(operation.operationId).to.be.ok;
        const camelCaseRe = /^(get|create|list|update|delete)([A-Z][a-z]+)+$/g;
        expect(
          camelCaseRe.test(operation.operationId || ''),
          `operationId "${
            operation.operationId || ''
          }" must be formatted [get|create|list|update|delete]camelCaseResource`,
        ).to.be.ok;
      },
    );
  },
  tags: ({ operations }: SnykApiCheckDsl) => {
    operations.requirement.must('have tags', (operation) => {
      expect(operation.tags).to.exist;
      expect(operation.tags).to.have.lengthOf.above(0, 'with examples');
    });
  },
  summary: ({ operations }: SnykApiCheckDsl) => {
    operations.requirement.must('have a summary', (operation) => {
      expect(operation.summary).to.exist;
    });
  },
  removingOperationId: ({ operations }: SnykApiCheckDsl) => {
    operations.changed.must(
      'have consistent operation IDs',
      (current, next) => {
        expect(current.operationId).to.equal(next.operationId);
      },
    );
  },
};
