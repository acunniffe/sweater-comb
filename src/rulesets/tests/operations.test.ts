import { rules } from "../operations";
import { createTestDslFixture } from "@useoptic/api-checks";
import { SnykApiCheckDsl, SynkApiCheckContext } from "../../dsl";

const { compare } = createTestDslFixture<SnykApiCheckDsl, SynkApiCheckContext>(
  (input) => {
    return new SnykApiCheckDsl(
      input.nextFacts,
      input.changelog,
      input.nextJsonLike,
      input.context
    );
  }
);

const emptyContext: SynkApiCheckContext = {
  changeDate: "2021-10-10",
  changeResource: "Example",
  changeVersion: {
    date: "2021-10-10",
    stability: "ga",
  },
  resourceVersions: {},
};

describe("operationId", () => {
  const baseForOperationIdTests = {
    openapi: "3.0.1",
    paths: {
      "/example": {
        get: {
          responses: {},
        },
      },
    },
    info: { version: "0.0.0", title: "Empty" },
  };

  describe("missing", () => {
    it("fails if empty string", async () => {
      const result = await compare(baseForOperationIdTests)
        .to((spec) => {
          spec.paths!["/example"]!.get!.operationId = "";
          return spec;
        })
        .withRule(rules.operationId, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it("fails if undefined", async () => {
      const result = await compare(baseForOperationIdTests)
        .to((spec) => {
          delete spec.paths!["/example"]!.get!.operationId;
          return spec;
        })
        .withRule(rules.operationId, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });

  describe("when set", () => {
    it("fails if prefix is wrong", async () => {
      const result = await compare(baseForOperationIdTests)
        .to((spec) => {
          spec.paths!["/example"]!.get!.operationId = "findHelloWorld";
          return spec;
        })
        .withRule(rules.operationId, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it("fails if not camel case", async () => {
      const result = await compare(baseForOperationIdTests)
        .to((spec) => {
          spec.paths!["/example"]!.get!.operationId = "get-hello-world";
          return spec;
        })
        .withRule(rules.operationId, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });

  it("fails if removed", async () => {
    const baseCopy = JSON.parse(JSON.stringify(baseForOperationIdTests));
    baseCopy.paths["/example"].get.operationId = "example";
    const result = await compare(baseCopy)
      .to((spec) => {
        delete spec.paths!["/example"]!.get!.operationId;
        return spec;
      })
      .withRule(rules.removingOperationId, emptyContext);

    expect(result.results[0].passed).toBeFalsy();
    expect(result).toMatchSnapshot();
  });

  it("fails if changed", async () => {
    // todo: fix copy/paste
    const baseCopy = JSON.parse(JSON.stringify(baseForOperationIdTests));
    baseCopy.paths["/example"].get.operationId = "example";
    const result = await compare(baseCopy)
      .to((spec) => {
        spec.paths!["/example"]!.get!.operationId = "example2";
        return spec;
      })
      .withRule(rules.removingOperationId, emptyContext);

    expect(result.results[0].passed).toBeFalsy();
    expect(result).toMatchSnapshot();
  });
});

const baseForOperationMetadataTests = {
  openapi: "3.0.1",
  paths: {
    "/example": {
      get: {
        tags: ["Example"],
        operationId: "getExample",
        summary: "Retrieve example",
        responses: {},
      },
    },
  },
  info: { version: "0.0.0", title: "Empty" },
};

describe("operation metadata", () => {
  describe("summary", () => {
    it("fails if missing", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => {
          delete spec.paths!["/example"]!.get!.summary;
          return spec;
        })
        .withRule(rules.summary, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it("passes if provided", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => {
          spec.paths!["/example"]!.get!.summary = "I have a summary";
          return spec;
        })
        .withRule(rules.summary, emptyContext);

      expect(result.results[0].passed).toBeTruthy();
      expect(result).toMatchSnapshot();
    });
  });

  describe("tags", () => {
    it("passes if > 1 tag provided", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => spec)
        .withRule(rules.tags, emptyContext);

      expect(result.results[0].passed).toBeTruthy();
      expect(result).toMatchSnapshot();
    });

    it("fail is not provided", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => {
          delete spec.paths!["/example"]!.get!.tags;
          return spec;
        })
        .withRule(rules.tags, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });
});

describe("operation parameters", () => {
  describe("names", () => {
    it("fails if the case isn't correct", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => {
          delete spec.paths!["/example"];
          spec.paths!["/example/{pathParameter}"] = {
            get: {
              parameters: [
                {
                  in: "path",
                  name: "pathParameter",
                },
              ],
              responses: {},
            },
          };
          return spec;
        })
        .withRule(rules.parameterCase, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it("passes if the case is correct", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => {
          delete spec.paths!["/example"];
          spec.paths!["/example/{path_parameter}"] = {
            get: {
              parameters: [
                {
                  in: "path",
                  name: "path_parameter",
                },
              ],
              responses: {},
            },
          };
          return spec;
        })
        .withRule(rules.parameterCase, emptyContext);

      expect(result.results[0].passed).toBeTruthy();
      expect(result).toMatchSnapshot();
    });
  });

  describe("version parameter", () => {
    it("fails when there is no version parameter", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => spec)
        .withRule(rules.versionParameter, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it("passes if there is a version parameter", async () => {
      const result = await compare(baseForOperationMetadataTests)
        .to((spec) => {
          spec.paths!["/example"]!.get!.parameters = [
            {
              in: "query",
              name: "version",
            },
          ];
          return spec;
        })
        .withRule(rules.parameterCase, emptyContext);

      expect(result.results[0].passed).toBeTruthy();
      expect(result).toMatchSnapshot();
    });
  });
});
