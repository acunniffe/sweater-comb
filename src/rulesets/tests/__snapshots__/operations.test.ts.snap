// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operation metadata summary fails if missing 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "tags": Array [
            "Example",
          ],
        },
        "before": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "tags": Array [
            "Example",
          ],
        },
      },
      "condition": "have a summary",
      "docsLink": undefined,
      "error": "expected undefined to exist",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;

exports[`operation metadata summary passes if provided 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "I have a summary",
          "tags": Array [
            "Example",
          ],
        },
        "before": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "I have a summary",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "I have a summary",
          "tags": Array [
            "Example",
          ],
        },
      },
      "condition": "have a summary",
      "docsLink": undefined,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;

exports[`operation metadata tags fail is not provided 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "Retrieve example",
        },
        "before": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "Retrieve example",
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "Retrieve example",
        },
      },
      "condition": "have tags",
      "docsLink": undefined,
      "error": "expected undefined to exist",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;

exports[`operation metadata tags passes if > 1 tag provided 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
      "condition": "have tags",
      "docsLink": undefined,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;

exports[`operationId fails if changed 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "example",
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "example2",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "operationId": "example",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "example2",
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changed": Object {
          "after": Object {
            "method": "get",
            "operationId": "example2",
            "pathPattern": "/example",
          },
          "before": Object {
            "method": "get",
            "operationId": "example",
            "pathPattern": "/example",
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
      },
      "condition": "have consistent operation IDs",
      "docsLink": undefined,
      "error": "expected 'example' to equal 'example2'",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated operation: GET /example",
    },
  ],
}
`;

exports[`operationId fails if removed 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "example",
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "operationId": "example",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changed": Object {
          "after": Object {
            "method": "get",
            "pathPattern": "/example",
          },
          "before": Object {
            "method": "get",
            "operationId": "example",
            "pathPattern": "/example",
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
      },
      "condition": "have consistent operation IDs",
      "docsLink": undefined,
      "error": "expected 'example' to equal undefined",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated operation: GET /example",
    },
  ],
}
`;

exports[`operationId missing fails if empty string 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "",
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "operationId": "",
          "pathPattern": "/example",
        },
      },
      "condition": "have the correct operationId format",
      "docsLink": undefined,
      "error": "expected '' to be truthy",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;

exports[`operationId missing fails if undefined 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "pathPattern": "/example",
        },
      },
      "condition": "have the correct operationId format",
      "docsLink": undefined,
      "error": "expected undefined to be truthy",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;

exports[`operationId when set fails if not camel case 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "get-hello-world",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "get-hello-world",
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "operationId": "get-hello-world",
          "pathPattern": "/example",
        },
      },
      "condition": "have the correct operationId format",
      "docsLink": undefined,
      "error": "operationId \\"get-hello-world\\" must be formatted [get|create|list|update|delete]camelCaseResource: expected false to be truthy",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;

exports[`operationId when set fails if prefix is wrong 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "findHelloWorld",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "findHelloWorld",
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "value": Object {
          "method": "get",
          "operationId": "findHelloWorld",
          "pathPattern": "/example",
        },
      },
      "condition": "have the correct operationId format",
      "docsLink": undefined,
      "error": "operationId \\"findHelloWorld\\" must be formatted [get|create|list|update|delete]camelCaseResource: expected false to be truthy",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for operation: GET /example",
    },
  ],
}
`;
